src/functional/ffront/past_passes/type_deduction.py

```diff
@@ -118,6 +120,47 @@ def visit_TupleExpr(self, node: past.TupleExpr, **kwargs):
             elts=elts, type=ct.TupleType(types=[el.type for el in elts]), location=node.location
         )
 
+    def _deduce_binop_type(
+        self,
+        node: past.BinOp,
+        *,
+        left: past.Expr,
+        right: past.Expr,
+        **kwargs,
+    ) -> Optional[ct.SymbolType]:
+        logical_ops = {ct.BinaryOperator.BIT_AND, ct.BinaryOperator.BIT_OR}
+        is_compatible = type_info.is_logical if node.op in logical_ops else type_info.is_arithmetic
+
+        # check both types compatible
+        for arg in (left, right):
+            if not isinstance(arg.type, ct.ScalarType) or not is_compatible(arg.type):
+                raise ProgramTypeError.from_past_node(
+                    arg, msg=f"Type {arg.type} can not be used in operator `{node.op}`!"
+                )
+
+        left_type = cast(ct.FieldType | ct.ScalarType, left.type)
+        right_type = cast(ct.FieldType | ct.ScalarType, right.type)
```

Comment:
```suggestion
        left_type = cast(ct.ScalarType, left.type)
        right_type = cast(ct.ScalarType, right.type)
```
Please check that this is still necessary or if mypy is able to deduce the type of `left` and `right` from the constraints in the loop above.